src/storage/leveldb/store.ts

```diff
@@ -0,0 +1,501 @@
+      fs.rmdirSync(this.dbpath, { recursive: true });
+    } else {
+      fs.rmSync(this.dbpath, { recursive: true, force: true });
+    }
+  }
+
+  async validate(): Promise<void> {
+    await this.addressIndex.validate();
+    await this.historyIndex.validate();
+    await this.utxoIndex.validate();
+    await this.tokenIndex.validate();
+    await this.walletIndex.validate();
+  }
+
+  async *addressIter(): AsyncGenerator<IAddressInfo> {
+    for await (const info of this.addressIndex.addressIter()) {
+      yield info;
+    }
+  }
+
+  async getAddress(base58: string): Promise<IAddressInfo | null> {
+    return this.addressIndex.getAddressInfo(base58);
+  }
+
+  async getAddressMeta(base58: string): Promise<IAddressMetadata | null> {
+    return this.addressIndex.getAddressMeta(base58);
+  }
+
+  async addressCount(): Promise<number> {
+    return this.addressIndex.addressCount();
+  }
+
+  async getAddressAtIndex(index: number): Promise<IAddressInfo | null> {
+    const address = await this.addressIndex.getAddressAtIndex(index);
+    if (address === null) {
+      return null;
+    }
+    return this.addressIndex.getAddressInfo(address);
+  }
+
+  async saveAddress(info: IAddressInfo): Promise<void> {
+    if (!info.base58) {
+      throw new Error('Invalid address');
+    }
+
+    if (await this.addressIndex.addressExists(info.base58)) {
+      throw new Error('Already have this address');
+    }
+
+    await this.addressIndex.saveAddress(info);
+
+    if ((await this.walletIndex.getCurrentAddressIndex()) === -1) {
+      await this.walletIndex.setCurrentAddressIndex(info.bip32AddressIndex);
+    }
+
+    if (info.bip32AddressIndex > (await this.walletIndex.getLastLoadedAddressIndex())) {
+      this.walletIndex.setLastLoadedAddressIndex(info.bip32AddressIndex);
+    }
+  }
+
+  async addressExists(base58: string): Promise<boolean> {
+    return this.addressIndex.addressExists(base58);
+  }
+
+  async getCurrentAddress(markAsUsed?: boolean | undefined): Promise<string> {
+    const addressIndex = await this.walletIndex.getCurrentAddressIndex();
+    const addressInfo = await this.getAddressAtIndex(addressIndex);
+    if (!addressInfo) {
+      throw new Error('Current address is not loaded');
+    }
+
+    if (markAsUsed) {
+      // Will move the address index only if we have not reached the gap limit
+      const lastLoadedIndex = await this.walletIndex.getLastLoadedAddressIndex();
+      await this.walletIndex.setCurrentAddressIndex(Math.min(lastLoadedIndex, addressIndex + 1));
+    }
+    return addressInfo.base58;
+  }
+
+  async *historyIter(tokenUid?: string | undefined): AsyncGenerator<IHistoryTx, any, unknown> {
+    for await (const tx of this.historyIndex.historyIter(tokenUid)) {
+      yield tx;
+    }
+  }
+
+  async historyCount(): Promise<number> {
+    return this.historyIndex.historyCount();
+  }
+
+  async processHistory({ rewardLock }: {rewardLock?: number} = {}): Promise<void> {
```

Comment:
This is a huge method, do you think we can (and should) split it into smaller methods?

Reply:
This is a huge method indeed, but it should not be split into smaller methods since it essentially is a single loop on the transaction history making aggregations and calculating metadata.